src/sparseml/transformers/sparsification/modification/modifying_llama.py

```diff
@@ -0,0 +1,251 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Modification to the original LLaMa model required in the
+context of SparseML
+"""
+
+import logging
+import math
+from typing import Optional, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+from transformers.models.llama.modeling_llama import (
+    LlamaAttention,
+    LlamaFlashAttention2,
+    apply_rotary_pos_emb,
+    repeat_kv,
+)
+
+from sparseml.pytorch.utils.helpers import swap_modules
+from sparseml.transformers.sparsification.modification.modification_objects import (
+    QuantizableIdentity,
+    QuantizableMatMul,
+)
+from sparseml.transformers.sparsification.modification.registry import (
+    ModificationRegistry,
+)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@ModificationRegistry.register(name="LlamaModel", alias=["LlamaForCausalLM"])
+def modify(model: nn.Module) -> nn.Module:
+    """
+    Modify the LLaMa model to be compatible with SparseML
+
+    1. Replaces the LlamaAttention modules with
+        LlamaAttentionWithQuantizableMatmuls modules
+
+    Note: This function will not alter any of the alternatives
+    to the LlamaAttention module such as LlamaFlashAttention2
+
+    :param model: the original LLaMa model
+    :return: the modified LLaMa model
+    """
+    for name, submodule in model.named_modules():
+        if isinstance(submodule, LlamaAttention):
+            swap_modules(model, name, LlamaAttentionWithQuantizableMatmuls(submodule))
+        elif isinstance(submodule, LlamaFlashAttention2):
+            _LOGGER.debug(
+                f"The model contains {submodule.__class__.__name__} "
+                "module, which will not be modified"
+            )
+    return model
+
+
+class MatMulLeftInput_QK(QuantizableIdentity):
+    ...
+
+
+class MatMulRightInput_QK(QuantizableIdentity):
+    ...
+
+
+class MatMulLeftInput_PV(QuantizableIdentity):
+    ...
+
+
+class MatMulRightInput_PV(QuantizableIdentity):
+    ...
+
+
+class MatMulOutput_QK(QuantizableIdentity):
+    ...
+
+
+class MatMulOutput_PV(QuantizableIdentity):
+    ...
```

Comment:
Maybe these could be pulled into the common quantized modules file? Since some of these are duplicates in mistral